### PR TITLE
ci(macos): add `MACOSX_DEPLOYMENT_TARGET` and friends

### LINKED ISSUE #3708: Rustup 1.27.0 incompatible with older macOS
### Problem

Trying to run `rustup self update` on macOS 10.14 fails with the error `dyld: cannot load 'rustup-init' (load command 0x80000034 is unknown)`.

Full output:
```
$ rustup self update
info: checking for self-update
info: downloading self-update
  rustup updated - (unknown) (from 1.26.0)

dyld: cannot load 'rustup-init' (load command 0x80000034 is unknown)
error: self-updated failed to replace rustup executable

$ uname -a
Darwin Caligari.local 18.7.0 Darwin Kernel Version 18.7.0: Tue Jun 22 19:37:08 PDT 2021; root:xnu-4903.278.70~1/RELEASE_X86_64 x86_64
```

### Steps

1. Run `rustup self update`
2. There is no step 2.

### Possible Solution(s)

_No response_

### Notes

_No response_

### Rustup version

```console
rustup 1.26.0 (5af9b9484 2023-04-05)
```

### Installed toolchains

```console
Default host: x86_64-apple-darwin
rustup home:  /Users/anders/.rustup

stable-x86_64-apple-darwin (default)
rustc 1.76.0 (07dca489a 2024-02-04)
```

### PR DESCRIPTION
**Pull request by rami3l**
Mirrors https://github.com/rust-lang/rust/blob/e61dcc7a0ac33ef054d76453307124233edcf545/src/ci/github-actions/ci.yml#L540-L552 and https://github.com/rust-lang/rust/blob/e61dcc7a0ac33ef054d76453307124233edcf545/src/ci/github-actions/ci.yml#L586-L605 by setting `build-macos-x86_64`'s target to macOS 10.12, and `build-macos-aarch64`'s target to macOS 11.0.

Fixes #3708.

(Am I doing this correctly, @ehuss?)

## Concerns

- [x] ~~Invite @andersm to download and test the CI build in order to verify the fix (as an ARM64 Mac user I don't get access to older macOS versions).~~ The feedback has been positive.

### CODE REVIEW

File: ci/actions-templates/macos-builds-template.yaml [hunk 1]
```diff
@@ -4,8 +4,14 @@ jobs: # skip-x86_64 skip-aarch64
...
   build-macos-aarch64: # job-name skip-x86_64
     runs-on: macos-14 # skip-x86_64
+    env: # skip-x86_64
+      MACOSX_DEPLOYMENT_TARGET: 11.0 # skip-x86_64
+      MACOSX_STD_DEPLOYMENT_TARGET: 11.0 # skip-x86_64
   build-macos-x86_64: # job-name skip-aarch64
     runs-on: macos-13 # skip-aarch64
+    env: # skip-aarch64
+      MACOSX_DEPLOYMENT_TARGET: 10.12 # skip-aarch64
+      MACOSX_STD_DEPLOYMENT_TARGET: 10.12 # skip-aarch64
```
Comment (ehuss) [inline ci/actions-templates/macos-builds-template.yaml]: `MACOSX_STD_DEPLOYMENT_TARGET` should not be set. That is something bootstrap uses when building the standard library.

Comment (rami3l) [reply ci/actions-templates/macos-builds-template.yaml]: @ehuss Thanks! I've fixed the code accordingly.

**Review by ehuss — APPROVED** [review]
Thanks! 👍

**Comment by rami3l** [comment]
@djc Can we merge this now? Thanks a lot :)

**Review by djc — APPROVED** [review]

**Pull request merged**